### PR TITLE
Restart CUDA 12 migration with CUDA arch builds

### PR DESCRIPTION
**Pull request by jakirkham**
As all CUDA Linux builds are now on equal footing, fold the CUDA 12 arch migration into the CUDA 12 migration and restart it. That way the latest migrator can be retrieved.

<hr>

Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by h-vetinari** [comment]
Isn't this still dependent on having had the arch migration on that feedstock first? Or have all CUDA feedstocks passed that migration already?

**Comment by jakirkham** [comment]
During [the core meeting on 2023 July 26]( https://conda-forge.org/docs/orga/minutes/2023-07-26.html ), we discussed how CUDA arch libraries make use of a newer GLIBC. In particular some libraries use symbols from GLIBC versions newer than 2.17. However GLIBC 2.28 support in conda-forge is still under discussion ( https://github.com/conda-forge/conda-forge.github.io/issues/1941 )

That said, one of the points brought up in the meeting is that from a conda-forge build perspective, mainly we are interested to know the GLIBC usage of static libraries that are linked in builds. With CUDA, this is `libcudart` generally. So a takeaway point from that meeting was to investigate symbols usage of `libcudart` and report back on what GLIBC was found

Have since done this and reported it in [the last conda-forge meeting]( https://conda-forge.org/docs/orga/minutes/2023-08-09.html ). Only saw GLIBC 2.17 symbols in use. Am including the results below for completeness:

```bash
$ conda create -n cuda cuda-cudart-dev -y
Channels:                                                                       
 - conda-forge
Platform: linux-aarch64
Collecting package metadata (repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/cuda

  added / updated specs:
    - cuda-cudart-dev


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    _openmp_mutex-4.5          |            2_gnu          23 KB  conda-forge
    cuda-cccl_linux-aarch64-12.0.90|       h579c4fd_1         1.1 MB  conda-forge
    cuda-cudart-12.0.107       |       hac28a21_6          21 KB  conda-forge
    cuda-cudart-dev-12.0.107   |       hac28a21_6          22 KB  conda-forge
    cuda-cudart-dev_linux-aarch64-12.0.107|       hac28a21_6         317 KB  conda-forge
    cuda-cudart-static-12.0.107|       hac28a21_6          22 KB  conda-forge
    cuda-cudart-static_linux-aarch64-12.0.107|       hac28a21_6         641 KB  conda-forge
    cuda-cudart_linux-aarch64-12.0.107|       hac28a21_6         180 KB  conda-forge
    cuda-version-12.0          |       hffde075_2          20 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.3 MB

The following NEW packages will be INSTALLED:

  _openmp_mutex      conda-forge/linux-aarch64::_openmp_mutex-4.5-2_gnu 
  cuda-cccl_linux-a~ conda-forge/noarch::cuda-cccl_linux-aarch64-12.0.90-h579c4fd_1 
  cuda-cudart        conda-forge/linux-aarch64::cuda-cudart-12.0.107-hac28a21_6 
  cuda-cudart-dev    conda-forge/linux-aarch64::cuda-cudart-dev-12.0.107-hac28a21_6 
  cuda-cudart-dev_l~ conda-forge/noarch::cuda-cudart-dev_linux-aarch64-12.0.107-hac28a21_6 
  cuda-cudart-static conda-forge/linux-aarch64::cuda-cudart-static-12.0.107-hac28a21_6 
  cuda-cudart-stati~ conda-forge/noarch::cuda-cudart-static_linux-aarch64-12.0.107-hac28a21_6 
  cuda-cudart_linux~ conda-forge/noarch::cuda-cudart_linux-aarch64-12.0.107-hac28a21_6 
  cuda-version       conda-forge/noarch::cuda-version-12.0-hffde075_2 
  libgcc-ng          conda-forge/linux-aarch64::libgcc-ng-13.1.0-h2b4548d_0 
  libgomp            conda-forge/linux-aarch64::libgomp-13.1.0-h2b4548d_0 
  libstdcxx-ng       conda-forge/linux-aarch64::libstdcxx-ng-13.1.0-h452befe_0 



Downloading and Extracting Packages
                                                                                
Preparing transaction: done                                                     
Verifying transaction: done                                                     
Executing transaction: done                                                     
#                                                                               
# To activate this environment, use                                             
#                                                                               
#     $ conda activate cuda                                                     
#                                                                               
# To deactivate an active environment, use
#
#     $ conda deactivate

$ strings /opt/conda/envs/cuda/targets/sbsa-linux/lib/libcudart_static.a | grep GLIBC
GLIBC_2.17
GLIBC_2.17
```

**Comment by jakirkham** [comment]
> Isn't this still dependent on having had the arch migration on that feedstock first? Or have all CUDA feedstocks passed that migration already?

If the arch migration has not occurred on a feedstock (either via the bot or a maintainer manually enabling those builds), then the CUDA 12 migrator will not add those jobs

If the the arch migration has run (or is run later), then the CUDA 12 arch builds will be added

IOW they are orthogonal

**Review by kkraus14 — APPROVED** [review]

**Comment by jakirkham** [comment]
@conda-forge/core, would be interested to hear your thoughts on this

File: recipe/migrations/cuda120.yaml
Comment (jakirkham) [inline recipe/migrations/cuda120.yaml:56]: Maybe it is worth calling out this change?

```suggestion
    Rebuild for CUDA 12 w/arch support
```

File: recipe/migrations/cuda120.yaml
Comment (jakirkham) [inline recipe/migrations/cuda120.yaml:6]: Need to double check this doesn't drop the migrator in migrated feedstocks

Comment (jakirkham) [reply recipe/migrations/cuda120.yaml:6]: IIUC [this code path in `conda-smithy`]( https://github.com/conda-forge/conda-smithy/blob/16666edc6c5d6b71fd205075c51edc0f408ab044/conda_smithy/configure_feedstock.py#L2245-L2253 ) continues to use the migrator already in a feedstock when a new `migration_number` occurs. IOW it will wait until the migrator bot add the new migration file (or a user adds it manually)

Removal doesn't occur as long as `migrator_ts` is the same, which is the case here

Comment (jakirkham) [reply recipe/migrations/cuda120.yaml:6]: Also IIUC the migrators used by `conda-smithy` [can be overridden by passing `--exclusive-config-file` during `rerender]( https://github.com/conda-forge/conda-smithy/blob/16666edc6c5d6b71fd205075c51edc0f408ab044/conda_smithy/configure_feedstock.py#L2220-L2226 ). So tried this with the [ucx-split-feedstock using latest changes]( https://github.com/conda-forge/ucx-split-feedstock/tree/0cf6eabee7753f9c526b6c679ca59492ce756e24 ) and found no change. So it looks like this is not an issue

**Comment by jakirkham** [comment]
Based on discussion in the conda-forge meeting today, it sounds like we are ok going ahead with this.

Only item was to check on the behavior of `migration_number`, which was done above (and behaves as we would intend).

Given this going to go ahead and get this started. Happy to follow up on anything else as needed. Thanks all! 🙏

**Comment by jakirkham** [comment]
Posted an update about this in the conda-forge CUDA 12 bringup issue: https://github.com/conda-forge/conda-forge.github.io/issues/1963#issuecomment-1690818611

File: recipe/migrations/cuda120.yaml
Comment (jakirkham) [reply recipe/migrations/cuda120.yaml:6]: Also have done a few re-renderings myself and with the bot after this was merged and everything seems to be working correctly

**Pull request merged**
